runtime-sdk/src/crypto/random.rs

```diff
@@ -37,37 +58,240 @@ impl Rng {
+        inner.transcript.append_message(b"tx", tx_hash.as_ref());
+    }
+
+    /// Append an observed subcontext to RNG transcript.
+    pub fn append_subcontext(&self) {
+        let mut inner = self.inner.borrow_mut();
+        inner.transcript.append_message(b"subctx", &[]);
+    }
+
+    /// Create an independent leaf RNG using this RNG as its parent.
+    pub fn fork<C: Context + ?Sized>(&self, ctx: &C, pers: &[u8]) -> Result<LeafRng, Error> {
+        let mut inner = self.inner.borrow_mut();
+
+        // Ensure the RNG is initialized and initialize it if not.
+        if inner.rng.is_none() {
+            // Derive the root VRF key for the current block.
+            let root_vrf_key = Self::derive_root_vrf_key(ctx)?;
+
+            // Initialize the root RNG.
+            let rng = root_vrf_key
+                .vrf_sign(&mut inner.transcript)
+                .0
+                .make_merlin_rng(&[]);
+            inner.rng = Some(rng);
+            // Reset transcript as we are initialized now.
+            inner.transcript = Transcript::new(b"oasis-runtime-sdk/crypto: rng post-init");
+        }
+
+        // Generate the leaf RNG.
+        let mut transcript = inner.transcript.clone();
+        transcript.append_message(b"fork", pers);
+
+        let parent_rng = inner.rng.as_mut().expect("rng must be initialized");
+        let rng = transcript.build_rng().finalize(parent_rng);
+
+        Ok(LeafRng(rng))
+    }
+}
+
+impl Default for RootRng {
+    fn default() -> Self {
+        Self::new()
     }
 }
 
-impl RngCore for Rng {
+/// A leaf RNG.
+pub struct LeafRng(TranscriptRng);
+
+impl RngCore for LeafRng {
     fn next_u32(&mut self) -> u32 {
-        self.rng.next_u32()
+        self.0.next_u32()
     }
 
     fn next_u64(&mut self) -> u64 {
-        self.rng.next_u64()
+        self.0.next_u64()
     }
 
     fn fill_bytes(&mut self, dest: &mut [u8]) {
-        self.rng.fill_bytes(dest)
+        self.0.fill_bytes(dest)
     }
 
     fn try_fill_bytes(&mut self, dest: &mut [u8]) -> Result<(), rand_core::Error> {
-        self.rng.try_fill_bytes(dest)
+        self.0.try_fill_bytes(dest)
     }
 }
 
-impl CryptoRng for Rng {}
+impl CryptoRng for LeafRng {}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use crate::{context::Mode, testing::mock};
+
+    #[test]
+    fn test_rng_basic() {
+        let mut mock = mock::Mock::default();
+        let ctx = mock.create_ctx_for_runtime::<mock::EmptyRuntime>(Mode::ExecuteTx, true);
+
+        // Create first root RNG.
+        let root_rng = RootRng::new();
+        let mut leaf_rng = root_rng.fork(&ctx, &[]).expect("rng fork should work");
+
+        let mut bytes1 = [0u8; 32];
+        leaf_rng.fill_bytes(&mut bytes1);
+
+        let mut leaf_rng = root_rng.fork(&ctx, &[]).expect("rng fork should work");
+        let mut bytes1_1 = [0u8; 32];
+        leaf_rng.fill_bytes(&mut bytes1_1);
+
+        assert_ne!(bytes1, bytes1_1, "rng should apply domain separation");
```

Comment:
oh because the root rng is stateful. leaf_rng that generated bytes1 is from the first use of root rng, leaf_rng that generated bytes1_1 is from second use

Reply:
Correct.